blocks/variables_dynamic.js

```diff
@@ -0,0 +1,162 @@
+  {
+    "type": "variables_set_dynamic",
+    "message0": "%{BKY_VARIABLES_SET}",
+    "args0": [{
+      "type": "field_variable",
+      "name": "VAR",
+      "variable": "%{BKY_VARIABLES_DEFAULT_NAME}"
+    },
+    {
+      "type": "input_value",
+      "name": "VALUE"
+    }
+    ],
+    "previousStatement": null,
+    "nextStatement": null,
+    "colour": "%{BKY_VARIABLES_DYNAMIC_HUE}",
+    "tooltip": "%{BKY_VARIABLES_SET_TOOLTIP}",
+    "helpUrl": "%{BKY_VARIABLES_SET_HELPURL}",
+    "extensions": ["contextMenu_variableDynamicSetterGetter"],
+    "mutator":'contextMenu_variableDynamicMutation'
+  }
+]); // END JSON EXTRACT (Do not delete this comment.)
+
+/**
+ * Mixin to add context menu items to create getter/setter blocks for this
+ * setter/getter.
+ * Used by blocks 'variables_set' and 'variables_get'.
+ * @mixin
+ * @augments Blockly.Block
+ * @package
+ * @readonly
+ */
+Blockly.Constants.VariablesDynamic.CUSTOM_CONTEXT_MENU_VARIABLE_GETTER_SETTER_MIXIN = {
+  /**
+     * Add menu option to create getter/setter block for this setter/getter.
+     * @param {!Array} options List of menu options to add to.
+     * @this Blockly.Block
+     */
+  customContextMenu: function(options) {
+    // Getter blocks have the option to create a setter block, and vice versa.
+    if(this.isInFlyout){
+      return;
+    }
+    var opposite_type ;
+    var contextMenuMsg ;
+    if (this.type == 'variables_get_dynamic') {
+      opposite_type = 'variables_set_dynamic';
+      contextMenuMsg = Blockly.Msg.VARIABLES_GET_CREATE_SET;
+    } else {
+      opposite_type = 'variables_get_dynamic';
+      contextMenuMsg = Blockly.Msg.VARIABLES_SET_CREATE_GET;
+    }
+
+    var option = { enabled: this.workspace.remainingCapacity() > 0 };
+    var name = this.getFieldValue('VAR');
+    option.text = contextMenuMsg.replace('%1', name);
+    var xmlField = goog.dom.createDom('field', null, name);
+    xmlField.setAttribute('name', 'VAR');
+    var variableModel = this.workspace.getVariable(name);
+    xmlField.setAttribute('variabletype', variableModel.type);
+    var xmlBlock = goog.dom.createDom('block', null, xmlField);
+    xmlBlock.setAttribute('type', opposite_type);
+    option.callback = Blockly.ContextMenu.callbackFactory(this, xmlBlock);
+    options.push(option);
+  },
+  onchange: function() {
+    var name = this.getFieldValue('VAR');
+    var variableModel = this.workspace.getVariable(name);
+    this.getField("VAR").variableTypes = [variableModel.type];
+    if (this.type == 'variables_get_dynamic') {
+      this.outputConnection.setCheck(variableModel.type);
+    } else {
+      this.getInput("VALUE").connection.setCheck(variableModel.type);
+    }
+  }
+};
+
+/**
+ * This mutator make sure the type check work before any event trigger.
+ * Some event handler check the type , and disconnect the connection which not match.
+ * @readonly
+ */
+Blockly.Constants.VariablesDynamic.CUSTOM_CONTEXT_MENU_VARIABLE_MUTATION = {
```

Comment:
I don't think you need a mutator for these blocks since the type checks don't need to be serialized. The change handler should update everything when the block is loaded from xml based on the variable.

Reply:
I think this block should know its' type anytime  since it has been created.
if I don't serialized the type infomation , it  can't get the type before some event triggered.
that will break something else when load the xml.

Reply:
I have tried extension . but the extension function runs before setValue , so that the block does not know its value.
if I do type checking in extension , it will disconnect the old connect first before setValue , and it will lose the old connection while setValue.
any suggest about this ?

Reply:
Oh, there's a bug in FieldVariable.getVariableTypes_ which was causing only empty type variables to get listed unless you specify the type. I did some quick tests and if you make the fix below this mutation can be removed and everything works fine.

Change core/field_variable.js line 142 from:
`if (variableTypes === null) {`
to:
`if (variableTypes === null || variableTypes === undefined) {`